lib/vector-core/src/event/mod.rs

```diff
@@ -435,13 +433,63 @@ impl MaybeAsLogMut for Event {
 
 /// A wrapper for references to inner event types, where reconstituting
 /// a full `Event` from a `LogEvent` or `Metric` might be inconvenient.
-#[derive(Clone, Copy, Debug)]
+#[derive(Clone, Copy, Debug, PartialEq)]
 pub enum EventRef<'a> {
     Log(&'a LogEvent),
     Metric(&'a Metric),
     Trace(&'a TraceEvent),
 }
 
+impl<'a> EventRef<'a> {
+    /// Extract the `LogEvent` reference in this.
+    ///
+    /// # Panics
+    ///
+    /// This will panic if this is not a `LogEvent` reference.
+    pub fn as_log(self) -> &'a LogEvent {
+        match self {
+            Self::Log(log) => log,
+            _ => panic!("Failed type coercion, {:?} is not a log reference", self),
+        }
+    }
+
+    /// Convert this reference into a new `LogEvent` by cloning.
+    ///
+    /// # Panics
+    ///
+    /// This will panic if this is not a `LogEvent` reference.
+    pub fn into_log(self) -> LogEvent {
+        match self {
+            Self::Log(log) => log.clone(),
+            _ => panic!("Failed type coercion, {:?} is not a log reference", self),
+        }
+    }
+
+    /// Extract the `Metric` reference in this.
+    ///
+    /// # Panics
+    ///
+    /// This will panic if this is not a `Metric` reference.
+    pub fn as_metric(self) -> &'a Metric {
+        match self {
+            Self::Metric(metric) => metric,
+            _ => panic!("Failed type coercion, {:?} is not a metric reference", self),
+        }
+    }
+
+    /// Convert this reference into a new `Metric` by cloning.
+    ///
+    /// # Panics
+    ///
+    /// This will panic if this is not a `Metric` reference.
+    pub fn into_metric(self) -> Metric {
+        match self {
+            Self::Metric(metric) => metric.clone(),
+            _ => panic!("Failed type coercion, {:?} is not a metric reference", self),
+        }
+    }
+}
```

Comment:
Missing as/into methods for traces?

Reply:
Well, nothing uses them yet, so I was deferring adding them until that point.

Reply:
I can accept that. 👍🏻